pkg/broker/ce_receiver.go

```diff
@@ -0,0 +1,331 @@
+}
+
+func (r *Receiver) dispatch(event *cloudevents.Event, uri *url.URL) (*cloudevents.Event, error) {
+	encodedEvent, err := r.codec.Encode(*event)
+	if err != nil {
+		return nil, err
+	}
+	msg, ok := encodedEvent.(*cehttp.Message)
+	if !ok {
+		return nil, errors.New("msg was not a cehttp.Message")
+	}
+
+	req, err := http.NewRequest(http.MethodPost, uri.String(), bytes.NewReader(msg.Body))
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header = msg.Header
+	res, err := r.httpClient.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	if res == nil {
+		// I don't think this is actually reachable with http.Client.Do(), but just to be sure we
+		// check anyway.
+		return nil, errors.New("non-error nil result from http.Client.Do()")
+	}
+
+	defer res.Body.Close()
+	if isFailure(res.StatusCode) {
+		// reject non-successful responses
+		return nil, fmt.Errorf("unexpected HTTP response, expected 2xx, got %d", res.StatusCode)
+	}
+
+	return r.decodeHTTPResponse(res)
+}
+
+// isFailure returns true if the status code is not a successful HTTP status.
+func isFailure(statusCode int) bool {
+	return statusCode < http.StatusOK /* 200 */ ||
+		statusCode >= http.StatusMultipleChoices /* 300 */
+}
+
+// Initialize the client. Mainly intended to load stuff in its cache.
+func (r *Receiver) initClient() error {
+	// We list triggers so that we can load the client's cache. Otherwise, on receiving an event, it
+	// may not find the trigger and would return an error.
+	opts := &client.ListOptions{
+		// Set Raw because if we need to get more than one page, then we will put the continue token
+		// into opts.Raw.Continue.
+		Raw: &metav1.ListOptions{},
+	}
+	for {
+		tl := &eventingv1alpha1.TriggerList{}
+		if err := r.client.List(context.TODO(), opts, tl); err != nil {
+			return err
+		}
+		if tl.Continue != "" {
+			opts.Raw.Continue = tl.Continue
+		} else {
+			break
+		}
+	}
+	return nil
+}
+
+func (r *Receiver) getTrigger(ctx context.Context, ref provisioners.ChannelReference) (*eventingv1alpha1.Trigger, error) {
+	t := &eventingv1alpha1.Trigger{}
+	err := r.client.Get(ctx,
+		types.NamespacedName{
+			Namespace: ref.Namespace,
+			Name:      ref.Name,
+		},
+		t)
+	return t, err
+}
+
+// shouldSendMessage determines whether message 'm' should be sent based on the triggerSpec 'ts'.
+// Currently it supports exact matching on type and/or source of events.
+func (r *Receiver) shouldSendMessage(ts *eventingv1alpha1.TriggerSpec, event *cloudevents.Event) bool {
+	if ts.Filter == nil || ts.Filter.SourceAndType == nil {
+		r.logger.Error("No filter specified")
+		return false
+	}
+	filterType := ts.Filter.SourceAndType.Type
+	if filterType != eventingv1alpha1.TriggerAnyFilter && filterType != event.Type() {
+		r.logger.Debug("Wrong type", zap.String("trigger.spec.filter.sourceAndType.type", filterType), zap.String("event.Type()", event.Type()))
+		return false
+	}
+	filterSource := ts.Filter.SourceAndType.Source
+	s := event.Context.AsV01().Source
```

Comment:
It seems that it would be nice to have a Source() function exposed in the cloud-events sdk.
If the event is actually v0.2, then this might try to read Ce-source instead of ce-source?